src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,124 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int NUM_ONE = 1;
+    private static final int DEFAULT_SIZE = 10;
+    private int size;
+    private T[] elementData;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        checkGrowIfNeed();
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAddByIndex(index);
+        checkGrowIfNeed();
+        arrayCopyForAddByIndex(index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (elementData.length > list.size()) {
+            for (int i = 0; i < list.size(); i++) {
+                elementData[size++] = list.get(i);
+            }
+        } else {
+            grow();
+            addAll(list);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = elementData[index];
+        removeByIndex(elementData, index);
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int index = 0; index < elementData.length; index++) {
+            if (equalsElemets(elementData[index], element)) {
+                removeByIndex(elementData, index);
+                return element;
+            }
+        }
+        throw new NoSuchElementException();
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkGrowIfNeed() {
+        if (size >= DEFAULT_SIZE) {
+            grow();
+        }
+    }
+
+    private T[] grow() {
+        int oldCapacity = elementData.length;
+        if (oldCapacity > 0 || elementData != null) {
```

Comment:
I'm not sure this 'if' is necessary here. Actually, if element.data is null, then elementData.length will cause NPE